### PR TITLE
[2.0] Merge up to 1.9

### PR DESCRIPTION
**Pull request by Jean85**
This is a continuation of the work of #526, continuosly porting all the new stuff from the master branch into the 2.0. This is the list of merged PRs on master that I've ported here (by merge order on master):

- [x] #543: docs fix
- [x] #546: adding `application/json` input handling
- [x] #547: docs fix bis
- [x] #548: fix Client PHPDoc
- [x] #551: fix syntax highlighting in the README.md
- [x] #556: docs fix ter
- [x] #557: use badge/poser in the README.md
- [x] ~~#555: Cleanup the 'site' value if it's empty~~ `site` is a deprecated field, no longer present
- [x] ~~#564: add runtime info~~ already covered by #562
- [x] #570: fix official docs definition
- [x] #572: handle non-default host port 
- [x] #573: prepare 1.8.4 release
- [x] ~~#578: Fix x-forwarded-proto detection~~ will be handled with #594
- [x] ~~#583: `exclude_exceptions` with child types~~ already handled in #487
- [x] #554: allow serialization 5 levels deep
- [x] #559: make string limit customizable
- [x] #587: avoid looping when exception throws exception
- [x] #586: Fix monolog handler not accepting Throwable
- [x] #553: fix serializer to account for non-UTF-8 chars
- [x] ~~#588: fix for undefined variable~~ not applicable, CurlHandler no longer present
- [x] #590: update for 1.9.0 release
- [x] ~~#591: prepare 1.10.x release~~ not applicable

I will handle those PHPT regression tests separately:
- [ ] #571: add regression tests for fatal handling
- [ ] #574: avoid double reporting
- [ ] #575: force send with curl_method async
- [ ] #576: more regression PHPT tests and fixes
- [ ] #589: fix regression test

### CODE REVIEW

**Comment by Jean85** [comment]
@ste93cry while trying to port #555 I discovered that the `site` field got lost during the development of the 2.x. Is it a custom field that we are ditching since we already have the standard `server_name`?

**Comment by ste93cry** [comment]
It was removed because it's a deprecated field

**Comment by Jean85** [comment]
@ste93cry looking at #578 I discovered that the `trust_x_forwarded_proto` option is handled but never used... what should we do?

**Comment by ste93cry** [comment]
According to the security issues that could arise by using the `X_FORWARDED_*` headers I would remove the configuration option entirely and not take any action regarding support of this particular feature in the PHP SDK

File: tests/SerializerAbstractTest.php [hunk 1]
```diff
@@ -409,6 +420,24 @@ public function testSetAllObjectSerialize()
...
         $this->assertFalse($serializer->getAllObjectSerialize());
     }
+
+    public function testClippingUTF8Characters()
+    {
+        if (!extension_loaded('mbstring')) {
+            $this->markTestSkipped('mbstring extension is not enabled.');
+        }
+
+        $testString = 'Прекратите надеяться, что ваши пользователи будут сообщать об ошибках';
```
Comment (Jean85) [inline tests/SerializerAbstractTest.php]: This is so funny! It's the Google-translated version of `Stop hoping that your users will report bugs` in russian :trollface:

**Comment by Jean85** [comment]
This is no longer WIP, I will handle the PHPT tests separately, please review!

**Comment by ste93cry** [comment]
I know that this argument has already been discussed many times and widely, but I strongly disagree with gathering the POST data and deserialize it if it's JSON. Now that we have a well organized way to collect all information we and users need through middlewares everyone can easily add whatever they want to the event data, and the SDK specs also says:

>This data should not be provided by default as it can get quite large

This is the main reason I did not collect POST data by default in the `RequestInterfaceMiddleware` middleware. If we really want to keep going this way, we should at least use the PSR-7 methods to get the data instead of using `file_get_contents` directly.

File: tests/SerializerAbstractTest.php (hunk 1 above)
Comment (stayallive) [reply tests/SerializerAbstractTest.php]: I couldn't just add some random text, needs to make a little bit sense at least 😉

**Comment by Jean85** [comment]
> This data should not be provided by default as it can get quite large

Oh well, then I think we should create a separated middleware to do that!

> This is the main reason I did not collect POST data by default in the RequestInterfaceMiddleware middleware. If we really want to keep going this way, we should at least use the PSR-7 methods to get the data instead of using file_get_contents directly.

What's the correct method to do it?

**Comment by ste93cry** [comment]
I think you should use the `getBody()` method of the `ServerRequestInterface` interface

**Comment by Jean85** [comment]
Ok I've reverted the commit about #546, I will handle it in a separate PR where I will create a separated middleware to handle POST body, with JSON handling too.

**Comment by ste93cry** [comment]
#587 should not apply to `2.x` as we don't set anymore the `event_id` property

**Comment by Jean85** [comment]
The `event_id` was set in the code of 1.x. The core of the issue is an exception that throws an other exception on `__set`, which would be triggered in a number of ways; in 1.x was the `event_id`.

**Comment by ste93cry** [comment]
What I meant is that I don't see anything in the test that tries to set the `event_id` property on the exception object (which in turn should throw another exception), so how can the test work? By the way, are you sure that the same case is not covered by my error handler PR?

**Comment by Jean85** [comment]
It's possible that is already covered by your PR, yes.

**Pull request merged**